### PR TITLE
【PRIM】Support use operator's output metadata info  in constructing static backward composite

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

Others
### PR changes

Others
### Describe

- support infer shape and dtype one-by-one operator in both python and c++ level
- support infer shape and dtype after append one grad operator
- replace shape,fill_const with fill_any_like for we can use static shape 
- fix relative magic number bugs
- add more tests

### CODE REVIEW

File: python/paddle/jit/dy2static/partial_program.py [hunk 1]
```diff
@@ -576,7 +576,7 @@ def _append_backward_desc(self, main_program):
                 core.check_and_set_prim_all_enabled()
             backward.gradients(targets=targets, inputs=[])
 
-        start_idx = len(main_program.block(0).ops) + 2 * len(
+        start_idx = len(main_program.block(0).ops) + 1 * len(
```
Comment (Aurelius84) [inline python/paddle/jit/dy2static/partial_program.py]:
```suggestion
        start_idx = len(main_program.block(0).ops) +  len(
```

File: python/paddle/jit/dy2static/partial_program.py [hunk 2]
```diff
@@ -753,7 +753,7 @@ def _get_forward_backward_program_form(
     ):
         # NOTE(dev): We apply build_strategy for backward firstly to
         # avoid skipping more gc variables.
-        backward_start_op_index = forward_end_op_index + 2 * len(
+        backward_start_op_index = forward_end_op_index + 1 * len(
```
Comment (Aurelius84) [inline python/paddle/jit/dy2static/partial_program.py]:
```suggestion
        backward_start_op_index = forward_end_op_index + len(
```

File: python/paddle/jit/dy2static/utils.py [hunk 3]
```diff
@@ -1512,12 +1512,11 @@ def _out_grad_names(program_desc, fwd_end_op_index, out_size):
     """
     names = []
     for i in range(
-        fwd_end_op_index + 1,
-        min(fwd_end_op_index + 2 * out_size, program_desc.block(0).op_size()),
-        2,
+        fwd_end_op_index,
+        min(fwd_end_op_index + 1 * out_size, program_desc.block(0).op_size()),
```
Comment (Aurelius84) [inline python/paddle/jit/dy2static/utils.py]:
```suggestion
        min(fwd_end_op_index + out_size, program_desc.block(0).op_size()),
```

File: python/paddle/jit/translated_layer.py [hunk 4]
```diff
@@ -373,7 +373,7 @@ def _forward_program_desc(self):
     @switch_to_static_graph
     def _create_backward_train_program(self):
         whole_program = _build_program_by_desc(self._train_program_desc)
-        start_op_index = self._infer_program_desc.block(0).op_size() + 2 * len(
+        start_op_index = self._infer_program_desc.block(0).op_size() + 1 * len(
```
Comment (Aurelius84) [inline python/paddle/jit/translated_layer.py]:
```suggestion
        start_op_index = self._infer_program_desc.block(0).op_size() + len(
```

File: python/paddle/fluid/tests/unittests/test_run_program_op.py [hunk 5]
```diff
@@ -131,7 +131,7 @@ def get_forward_backward_program_desc(
         forward_program = _add_build_strategy_for(program, 0, forward_op_num)
         backward_program = _add_build_strategy_for(
             program,
-            forward_op_num + 2 * output_num,
+            forward_op_num + 1 * output_num,
```
Comment (Aurelius84) [inline python/paddle/fluid/tests/unittests/test_run_program_op.py]:
```suggestion
            forward_op_num + output_num,
```

File: python/paddle/jit/dy2static/partial_program.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/jit/dy2static/partial_program.py]: done

File: python/paddle/jit/dy2static/partial_program.py (hunk 2 above)
Comment (cxxly) [reply python/paddle/jit/dy2static/partial_program.py]: done

File: python/paddle/jit/dy2static/utils.py (hunk 3 above)
Comment (cxxly) [reply python/paddle/jit/dy2static/utils.py]: done

File: python/paddle/jit/translated_layer.py (hunk 4 above)
Comment (cxxly) [reply python/paddle/jit/translated_layer.py]: done

File: python/paddle/fluid/tests/unittests/test_run_program_op.py (hunk 5 above)
Comment (cxxly) [reply python/paddle/fluid/tests/unittests/test_run_program_op.py]: done

File: paddle/fluid/prim/utils/static/static_global_utils.h [hunk 6]
```diff
@@ -69,12 +69,21 @@ class StaticCompositeContext {
     enable_bwd_prim_ = enable_prim;
   }
 
+  void SetTargetGradName(const std::map<std::string, std::string>& m) {
+    target_grad_name_ = m;
+  }
+
+  std::map<std::string, std::string> GetTargetGradName() {
```
Comment (JiabinYang) [inline paddle/fluid/prim/utils/static/static_global_utils.h:76]: Maybe just give a check method instead of return whole map

**Review by JiabinYang — COMMENTED** [review]
one comment

File: paddle/fluid/prim/utils/static/static_global_utils.h (hunk 6 above)
Comment (cxxly) [reply paddle/fluid/prim/utils/static/static_global_utils.h:76]: fixed in a new pr.

**Review by JiabinYang — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]

**Pull request merged**